transport/tlscommon/tls_config_test.go

```diff
@@ -678,30 +676,111 @@ func startTestServer(t *testing.T, serverAddr string, serverCerts []tls.Certific
 	return *serverURL
 }
 
-func openTestCerts(t testing.TB) map[string]*x509.Certificate {
-	t.Helper()
-	certs := make(map[string]*x509.Certificate, 0)
+func getFingerprint(cert *x509.Certificate) string {
+	caSHA256 := sha256.Sum256(cert.Raw)
+	return hex.EncodeToString(caSHA256[:])
+}
 
-	for testcase, certname := range map[string]string{
-		"expired":           "tls.crt",
-		"unknown authority": "unsigned_tls.crt",
-		"correct":           "client1.crt",
-		"wildcard":          "server.crt",
-		"es-leaf":           "es-leaf.crt",
-		"es-root-ca":        "es-root-ca-cert.crt",
-	} {
+func genTestCerts(t *testing.T) map[string]*x509.Certificate {
+	ca, err := genCA()
+	if err != nil {
+		t.Fatalf("cannot generate root CA: %s", err)
+	}
 
-		certBytes, err := ioutil.ReadFile(filepath.Join("testdata", certname))
-		if err != nil {
-			t.Fatalf("reading file %q: %+v", certname, err)
-		}
-		block, _ := pem.Decode(certBytes)
-		testCert, err := x509.ParseCertificate(block.Bytes)
+	unknownCA, err := genCA()
+	if err != nil {
+		t.Fatalf("cannot generate second root CA: %s", err)
+	}
+
+	certs := map[string]*x509.Certificate{
+		"ca": ca.Leaf,
+	}
+
+	certData := map[string]struct {
+		ca       tls.Certificate
+		keyUsage x509.KeyUsage
+		isCA     bool
+		dnsNames []string
+		ips      []net.IP
+		expired  bool
+	}{
+		"wildcard": {
+			ca:       ca,
+			keyUsage: x509.KeyUsageDigitalSignature,
+			isCA:     false,
+			dnsNames: []string{"*.example.com"},
+		},
+		"correct": {
+			ca:       ca,
+			keyUsage: x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+			isCA:     false,
+			dnsNames: []string{"localhost"},
+			// IPV4 and IPV6
+			ips: []net.IP{{127, 0, 0, 1}, {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}},
+		},
+		"unknown authority": {
+			ca:       unknownCA,
+			keyUsage: x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+			isCA:     false,
+			dnsNames: []string{"localhost"},
+			// IPV4 and IPV6
+			ips: []net.IP{{127, 0, 0, 1}, {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}},
+		},
+		"expired": {
+			ca:       ca,
+			keyUsage: x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+			isCA:     false,
+			dnsNames: []string{"localhost"},
+			// IPV4 and IPV6
+			ips:     []net.IP{{127, 0, 0, 1}, {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}},
+			expired: true,
+		},
+	}
+
+	for certName, data := range certData {
+		cert, err := genSignedCert(
+			data.ca,
+			data.keyUsage,
+			data.isCA,
+			certName,
+			data.dnsNames,
+			data.ips,
+			data.expired,
+		)
 		if err != nil {
-			t.Fatalf("parsing certificate %q: %+v", certname, err)
+			t.Fatal(err)
 		}
-		certs[testcase] = testCert
+		certs[certName] = cert.Leaf
 	}
 
+	// If for any reason there is a need to debug
```

Comment:
If for some reason this fails in CI, we are going to want the cert/key pair to see what happened.  How about always dumping out the keys & certs, then deleting the dir if the test succeeds?

Reply:
Yeah, I first thought about it, but while writing this PR what helped me was to have a fixed set of certs that I could use over multiple test runs. So I ended up generating them once and re-using until I found out where the mismatch between cert, CA and fingerprint were.

Anyway, I see no issue in persisting them on failure.